File: worker/worker/trust.py
"""Trust Score + Risk Flags — pure computation, no I/O.

Scoring is intentionally strict. A high trust score must be *earned*
across multiple independent signals — popularity alone is not enough.

Dimensions (each 0-100, then weighted):
  activity_score    (25%) — recency of pushes/commits, penalises stale repos
  community_score   (20%) — fork-to-star health, watchers, contributor signal
  maintenance_score (25%) — issue ratio, recency, description, topics, license
  popularity_score  (15%) — stars with aggressive diminishing returns
  maturity_score    (15%) — repo age, consistent history, not brand-new
"""

import math
from datetime import datetime, timedelta, timezone


def _clamp(val: float, lo: float = 0.0, hi: float = 100.0) -> float:
    return max(lo, min(hi, val))


def _now_utc() -> datetime:
    return datetime.now(timezone.utc)


def _make_aware(dt: datetime) -> datetime:
    """Ensure a datetime is timezone-aware (UTC)."""
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt


def _days_since(dt: datetime | str | None) -> float:
    if dt is None:
        return 365 * 5  # treat unknown as very old
    if isinstance(dt, str):
        try:
            dt = datetime.fromisoformat(dt)
        except Exception:
            return 365 * 5
    delta = _now_utc() - _make_aware(dt)
    return max(delta.total_seconds() / 86400, 0)


# ── Dimension scorers (public for individual use) ──────────

def compute_activity_score(last_pushed_at, last_commit_at) -> float:
    """More recent = higher score. Caps at 95 — perfection is suspicious."""
    best = min(_days_since(last_pushed_at), _days_since(last_commit_at))
    if best <= 7:
        return 95.0
    if best <= 30:
        return 85.0
    if best <= 90:
        return 65.0
    if best <= 180:
        return 45.0
    if best <= 365:
        return 25.0
    return max(5.0, 25.0 - (best - 365) / 25)


def compute_community_score(stars: int, forks: int, watchers: int) -> float:
    """Healthy community = meaningful forks relative to stars + watchers."""
    f = _clamp(math.log2(forks + 1) * 8, 0, 70)
    w = _clamp(math.log2(watchers + 1) * 10, 0, 60)

    # Fork-to-star health
    if stars > 50:
        ratio = forks / max(stars, 1)
        health = _clamp(ratio * 300, 0, 100)
    else:
        health = 50.0

    return round(f * 0.35 + w * 0.25 + health * 0.40, 2)


def compute_maintenance_score(
    open_issues: int,
    stars: int,
    last_pushed_at,
    license_name: str | None = None,
    description: str | None = None,
    has_topics: bool = False,
) -> float:
    """Multi-factor maintenance: issue ratio, recency, license, description, topics."""
    if stars == 0:
        ratio_score = 40.0
    else:
        ratio = open_issues / stars
        ratio_score = _clamp(100 - ratio * 600, 0, 100)

    recency = compute_activity_score(last_pushed_at, None) * 0.4

    license_score = 80.0 if license_name else 10.0

    desc_score = 60.0
    if description and len(description.strip()) >= 30:
        desc_score = 90.0
    elif description and len(description.strip()) >= 10:
        desc_score = 70.0
    elif not description or len(description.strip()) < 3:
        desc_score = 20.0

    topic_score = 75.0 if has_topics else 35.0

    return round(
        ratio_score * 0.30
        + recency * 0.25
        + license_score * 0.25
        + desc_score * 0.10
        + topic_score * 0.10,
        2,
    )


def compute_popularity_score(stars: int) -> float:
    """Aggressive diminishing returns — 10k stars ≈ 80, not 100."""
    if stars <= 0:
        return 0.0
    return _clamp(math.log10(stars) * 20, 0, 85)


def compute_maturity_score(created_at, last_pushed_at, stars: int, forks: int) -> float:
    """Older repos with sustained activity score higher."""
    age_days = _days_since(created_at)
    active_days = _days_since(last_pushed_at)

    if age_days < 30:
        age_score = 10.0
    elif age_days < 90:
        age_score = 25.0
    elif age_days < 365:
        age_score = 50.0
    elif age_days < 730:
        age_score = 70.0
    else:
        age_score = min(90.0, 70.0 + math.log2(age_days / 730) * 10)

    if age_days > 365 and active_days < 180:
        sustained_bonus = 20.0
    elif age_days > 180 and active_days < 90:
        sustained_bonus = 10.0
    else:
        sustained_bonus = 0.0

    if stars < 3 and forks < 1:
        age_score *= 0.5

    return _clamp(age_score + sustained_bonus, 0, 100)


# ── Combined trust score ───────────────────────────────────

def compute_trust_score(
    stars: int,
    forks: int,
    watchers: int,
    open_issues: int,
    last_pushed_at,
    last_commit_at,
    *,
    license_name: str | None = None,
    description: str | None = None,
    has_topics: bool = False,
    created_at=None,
) -> dict:
    activity = round(compute_activity_score(last_pushed_at, last_commit_at), 2)
    community = round(compute_community_score(stars, forks, watchers), 2)
    maintenance = round(compute_maintenance_score(open_issues, stars, last_pushed_at, license_name, description, has_topics), 2)
    popularity = round(compute_popularity_score(stars), 2)
    maturity = round(compute_maturity_score(created_at, last_pushed_at, stars, forks), 2)

    overall = round(
        activity * 0.25
        + community * 0.20
        + maintenance * 0.25
        + popularity * 0.15
        + maturity * 0.15,
        2,
    )

    # Hard ceilings
    if not license_name and overall > 60:
        overall = 60.0
    if stars < 5 and overall > 50:
        overall = 50.0
    if forks < 1 and overall > 55:
        overall = 55.0

    return {
        "overall": round(overall, 2),
        "activity_score": activity,
        "community_score": community,
        "maintenance_score": maintenance,
        "popularity_score": popularity,
        "maturity_score": maturity,
    }


def compute_risk_flags(
    stars: int,
    forks: int,
    open_issues: int,
    license_name: str | None,
    last_pushed_at,
    *,
    description: str | None = None,
    created_at=None,
) -> list[dict]:
    flags = []
    days = _days_since(last_pushed_at)
    age_days = _days_since(created_at)

    if not license_name:
        flags.append({
            "flag_type": "no_license",
            "severity": "high",
            "message": "No license detected — usage rights unclear.",
        })

    if days > 365:
        flags.append({
            "flag_type": "stale",
            "severity": "high" if days > 730 else "medium",
            "message": f"Last activity was {int(days)} days ago.",
        })

    if forks < 2 and stars < 10:
        flags.append({
            "flag_type": "low_adoption",
            "severity": "medium",
            "message": "Very few forks and stars — low community adoption.",
        })

    if open_issues > stars * 0.5 and stars > 0:
        flags.append({
            "flag_type": "high_issue_ratio",
            "severity": "medium",
            "message": f"Open issues ({open_issues}) are >50 % of stars ({stars}).",
        })

    if age_days < 90:
        flags.append({
            "flag_type": "very_new",
            "severity": "medium",
            "message": f"Repository is only {int(age_days)} days old — limited track record.",
        })

    if not description or len(description.strip()) < 10:
        flags.append({
            "flag_type": "no_description",
            "severity": "low",
            "message": "Missing or very short description.",
        })

    if stars > 100 and forks < stars * 0.01:
        flags.append({
            "flag_type": "suspicious_stars",
            "severity": "medium",
            "message": f"Star-to-fork ratio is unusually low ({forks} forks / {stars} stars).",
        })

    return flags

File: frontend/src/lib/trust.ts
/**
 * Trust Score + Risk Flags — pure computation in TypeScript.
 * Port of backend/app/trust.py
 *
 * Scoring is intentionally strict. A high trust score must be *earned*
 * across multiple independent signals — popularity alone is not enough.
 *
 * Dimensions (each 0-100, then weighted):
 *   activity_score    (25%) — recency of pushes/commits
 *   community_score   (20%) — fork-to-star health, watchers
 *   maintenance_score (25%) — issue ratio, recency, license, desc, topics
 *   popularity_score  (15%) — stars with aggressive diminishing returns
 *   maturity_score    (15%) — repo age, sustained history
 */

export interface TrustScore {
  overall: number;
  activity_score: number;
  community_score: number;
  maintenance_score: number;
  popularity_score: number;
  maturity_score: number;
}

export interface RiskFlag {
  flag_type: string;
  severity: string;
  message: string;
}

function clamp(val: number, lo = 0, hi = 100): number {
  return Math.max(lo, Math.min(hi, val));
}

function daysSince(dateStr: string | null | undefined): number {
  if (!dateStr) return 365 * 5;
  try {
    const diff = Date.now() - new Date(dateStr).getTime();
    return Math.max(diff / 86400_000, 0);
  } catch {
    return 365 * 5;
  }
}

function r2(v: number): number {
  return Math.round(v * 100) / 100;
}

// ── Dimension scorers ───────────────────────────────────────

function activityScore(lastPushed: string | null, lastCommit: string | null): number {
  const best = Math.min(daysSince(lastPushed), daysSince(lastCommit));
  if (best <= 7) return 95;
  if (best <= 30) return 85;
  if (best <= 90) return 65;
  if (best <= 180) return 45;
  if (best <= 365) return 25;
  return Math.max(5, 25 - (best - 365) / 25);
}

function communityScore(stars: number, forks: number, watchers: number): number {
  const f = clamp(Math.log2(forks + 1) * 8, 0, 70);
  const w = clamp(Math.log2(watchers + 1) * 10, 0, 60);

  let health: number;
  if (stars > 50) {
    const ratio = forks / Math.max(stars, 1);
    health = clamp(ratio * 300, 0, 100);
  } else {
    health = 50;
  }

  return r2(f * 0.35 + w * 0.25 + health * 0.40);
}

function maintenanceScore(
  openIssues: number,
  stars: number,
  lastPushed: string | null,
  licenseName: string | null,
  description: string | null,
  hasTopics: boolean,
): number {
  const ratioScore = stars === 0 ? 40 : clamp(100 - (openIssues / stars) * 600, 0, 100);
  const recency = activityScore(lastPushed, null) * 0.4;
  const licenseScore = licenseName ? 80 : 10;

  let descScore = 60;
  if (description && description.trim().length >= 30) descScore = 90;
  else if (description && description.trim().length >= 10) descScore = 70;
  else if (!description || description.trim().length < 3) descScore = 20;

  const topicScore = hasTopics ? 75 : 35;

  return r2(
    ratioScore * 0.30 +
    recency * 0.25 +
    licenseScore * 0.25 +
    descScore * 0.10 +
    topicScore * 0.10,
  );
}

function popularityScore(stars: number): number {
  if (stars <= 0) return 0;
  return clamp(Math.log10(stars) * 20, 0, 85);
}

function maturityScore(
  createdAt: string | null,
  lastPushed: string | null,
  stars: number,
  forks: number,
): number {
  const ageDays = daysSince(createdAt);
  const activeDays = daysSince(lastPushed);

  let ageScore: number;
  if (ageDays < 30) ageScore = 10;
  else if (ageDays < 90) ageScore = 25;
  else if (ageDays < 365) ageScore = 50;
  else if (ageDays < 730) ageScore = 70;
  else ageScore = Math.min(90, 70 + Math.log2(ageDays / 730) * 10);

  let sustainedBonus = 0;
  if (ageDays > 365 && activeDays < 180) sustainedBonus = 20;
  else if (ageDays > 180 && activeDays < 90) sustainedBonus = 10;

  if (stars < 3 && forks < 1) ageScore *= 0.5;

  return clamp(ageScore + sustainedBonus, 0, 100);
}

// ── Combined trust score ────────────────────────────────────

export function computeTrustScore(
  stars: number,
  forks: number,
  watchers: number,
  openIssues: number,
  lastPushed: string | null,
  lastCommit: string | null,
  opts: {
    licenseName?: string | null;
    description?: string | null;
    hasTopics?: boolean;
    createdAt?: string | null;
  } = {},
): TrustScore {
  const licenseName = opts.licenseName ?? null;
  const description = opts.description ?? null;
  const hasTopics = opts.hasTopics ?? false;
  const createdAt = opts.createdAt ?? null;

  const activity = r2(activityScore(lastPushed, lastCommit));
  const community = communityScore(stars, forks, watchers);
  const maintenance = maintenanceScore(openIssues, stars, lastPushed, licenseName, description, hasTopics);
  const popularity = r2(popularityScore(stars));
  const maturity = r2(maturityScore(createdAt, lastPushed, stars, forks));

  let overall = r2(
    activity * 0.25 +
    community * 0.20 +
    maintenance * 0.25 +
    popularity * 0.15 +
    maturity * 0.15,
  );

  // Hard ceilings — can't bluff past these
  if (!licenseName && overall > 60) overall = 60;
  if (stars < 5 && overall > 50) overall = 50;
  if (forks < 1 && overall > 55) overall = 55;

  return {
    overall: r2(overall),
    activity_score: activity,
    community_score: community,
    maintenance_score: maintenance,
    popularity_score: popularity,
    maturity_score: maturity,
  };
}

export function computeRiskFlags(
  stars: number,
  forks: number,
  openIssues: number,
  licenseName: string | null,
  lastPushed: string | null,
  opts: {
    description?: string | null;
    createdAt?: string | null;
  } = {},
): RiskFlag[] {
  const flags: RiskFlag[] = [];
  const days = daysSince(lastPushed);
  const ageDays = daysSince(opts.createdAt);

  if (!licenseName) {
    flags.push({
      flag_type: 'no_license',
      severity: 'high',
      message: 'No license detected — usage rights unclear.',
    });
  }

  if (days > 365) {
    flags.push({
      flag_type: 'stale',
      severity: days > 730 ? 'high' : 'medium',
      message: `Last activity was ${Math.floor(days)} days ago.`,
    });
  }

  if (forks < 2 && stars < 10) {
    flags.push({
      flag_type: 'low_adoption',
      severity: 'medium',
      message: 'Very few forks and stars — low community adoption.',
    });
  }

  if (openIssues > stars * 0.5 && stars > 0) {
    flags.push({
      flag_type: 'high_issue_ratio',
      severity: 'medium',
      message: `Open issues (${openIssues}) are >50% of stars (${stars}).`,
    });
  }

  if (ageDays < 90) {
    flags.push({
      flag_type: 'very_new',
      severity: 'medium',
      message: `Repository is only ${Math.floor(ageDays)} days old — limited track record.`,
    });
  }

  if (!opts.description || opts.description.trim().length < 10) {
    flags.push({
      flag_type: 'no_description',
      severity: 'low',
      message: 'Missing or very short description.',
    });
  }

  if (stars > 100 && forks < stars * 0.01) {
    flags.push({
      flag_type: 'suspicious_stars',
      severity: 'medium',
      message: `Star-to-fork ratio is unusually low (${forks} forks / ${stars} stars).`,
    });
  }

  return flags;
}
